lighthouse-core/audits/non-composited-animations.js

```diff
@@ -0,0 +1,178 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const TraceOfTab = require('../computed/trace-of-tab.js');
+const i18n = require('../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Title of a diagnostic LH audit that provides details on animations that are not composited. */
+  title: 'Avoid non-composited animations',
+  /** Description of a diagnostic LH audit that shows the user animations that are not composited. */
+  description: 'Animations which are not composited can be janky and contribute to CLS. ' +
+    '[Learn more]()',
+  /** [ICU Syntax] Label identifying the number of animations that are not composited. */
+  displayValue: `{itemCount, plural,
+  =1 {# animation found}
+  other {# animations found}
+  }`,
+  /** Label for when the animation fails to composite due to an unsupported CSS property. */
+  unsupportedCSS: 'Unsupported CSS Property',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @type {{flag: number, text: string}[]} */
+const ACTIONABLE_FAILURE_REASONS = [
+  {
+    flag: 1 << 13,
+    text: UIStrings.unsupportedCSS,
+  },
+];
+
+/**
+ * Return list of actionable failure reasons and a boolean if some reasons are not actionable.
+ * @param {number} failureCode
+ * @return {{failureReasons: string[], hasNonActionable: boolean}}
+ */
+function getActionableFailureReasons(failureCode) {
+  /** @type {string[]} */
+  const failureReasons = [];
+  ACTIONABLE_FAILURE_REASONS.forEach(reason => {
+    if (failureCode & reason.flag) {
+      failureReasons.push(reason.text);
+      failureCode &= ~reason.flag;
+    }
+  });
+  return {failureReasons, hasNonActionable: Boolean(failureCode)};
+}
+
+class NonCompositedAnimations extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'non-composited-animations',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: str_(UIStrings.title),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'AnimatedElements'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const traceOfTab = await TraceOfTab.request(trace, context);
+
+    /** @type {Map<string, {begin: LH.TraceEvent | undefined, status: LH.TraceEvent | undefined}>} */
+    const animationPairs = new Map();
+    traceOfTab.mainThreadEvents.forEach(event => {
```

Comment:
it's not a strong preference and we have plenty of `forEach` in the codebase too, so I wouldn't worry about this normally but in the case of iterating through every trace event in a large block, I personally find it worthwhile to get in the habit of `for (const event of events)` for the style match and [free 10x perf boost](https://jsperf.com/for-vs-forin-vs-forof-vs-foreach). helps make sure we don't end up with a death by a thousand cuts too :)

(we don't comment on perf optimizations until they would be significant, but first PR iterating through all trace events is worth it 😃  at least for the future)